### PR TITLE
LayerSwitcher_#561 improvement

### PR DESCRIPTION
**Pull request by martyrion**
New Layer Switcher:

- movable window
- option to create without a switcher (button)
- option to have the sections expanded
- minimize/maximize window
- expand/collapse sections
- vertical resize window
- double click header to reset original position

#### **Verify that...**

- [x] `npm test` passes
- [x] `npm run api` passes
- [x] `npm run lint` passes
- [x] `npm run build` still works

### CODE REVIEW

**Comment by martyrion** [comment]
I have kept the old LayerSwitcher and name the new one **_LayerSwitcherv2_**
Also this breaks functionality with other switchers. If you approve it I can change the other switchers too.

File: css/og.css [hunk 1]
```diff
@@ -200,81 +221,178 @@
...
+    height: 24px;
+    width: 24px;
+    cursor: pointer;
+}
+
+/* All buttons have a div inside to hold the background base64 image */
+.og-dialog-header-btn>div {
+    width: 24px;
+    height: 24px;
+    background-size: 14px !important;
```
Comment (Zemledelec) [inline css/og.css]: Is it possible to remove !important?

File: src/og/control/LayerSwitcherv2.js [hunk 2]
```diff
@@ -0,0 +1,490 @@
...
+        const layerEvents = () => {
+            this.planet.events.on("layeradd", this.addNewLayer, this)
+            this.planet.events.on("layerremove", this.removeLayer, this)
+        }
+
+        this.planet.events.on("rendercompleted", layerEvents, this)
+    }
+
+    // BASIC DATA COLLECTION-PREPARATION FUNCTIONS
+    planetDataBasic = () => {
```
Comment (Zemledelec) [inline src/og/control/LayerSwitcherv2.js]: No need  methods to be arrow functions

File: src/og/control/LayerSwitcherv2.js [hunk 3]
```diff
@@ -0,0 +1,490 @@
...
+            else { return 'outside' }
+        }
+
+        const whereClickHandler = (e) => {
+            // Check where I clicked : in dialog, in button, elsewhere
+            let whereIclicked = D_menu_btn ? whereClick(e, D_dialog, D_menu_btn) : null
+
+            // If I clicked elsewhere and the dialog is unpinned, then hide dialog and set menu btn to OFF
+            if (whereIclicked == 'outside' && D_header_pin.dataset.attachement == 'UNPINNED') {
+                D_dialog.classList.add('og-not-visible')
```
Comment (Zemledelec) [inline src/og/control/LayerSwitcherv2.js]: In the openglobus js is only camelCase style in the code. Rename variables, please.

File: src/og/control/LayerSwitcherv2.js [hunk 4]
```diff
@@ -0,0 +1,490 @@
...
+                D_menu_btn.classList.add('og-OFF')
```
Comment (Zemledelec) [inline src/og/control/LayerSwitcherv2.js]: HTMLElement variables must be started from $, for instance $dialogHeaderPin

File: src/og/control/LayerSwitcherv2.js [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+                // If Iclicked on button, toggle dialog and menu btn
+            } else if (whereIclicked == 'on-btn') {
```
Comment (Zemledelec) [inline src/og/control/LayerSwitcherv2.js]: use ===

File: src/og/control/LayerSwitcherv2.js [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+                D_dialog.classList.toggle('og-not-visible')
+                D_menu_btn.classList.toggle('og-OFF')
+            }
+        }
+
+        // Document
+        document.addEventListener('click', (e) => whereClickHandler(e))
```
Comment (Zemledelec) [inline src/og/control/LayerSwitcherv2.js]: If the layer switcher is removed there will be a leack. It must be resolved with removeEventListener in control deactivation virtual method.

File: src/og/control/LayerSwitcherv2.js [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+                return elementFactory('details', { class: 'og-layer-switcher-record-title' },
+                    elementFactory('summary', {}, object.name || object.url || 'no-name'))
+            } else {
+                return elementFactory('label', { class: 'og-layer-switcher-record-title' },
+                    object.name || object.properties.name || 'no-name')
+            }
+        }
+
+        const visibility = (object, nameConcat) => {
+            if (nameConcat == 'TerrainProviders' && planet.terrain == object) {
```
Comment (Zemledelec) [inline src/og/control/LayerSwitcherv2.js]: maybe it is better to refactor these cases into one?

File: src/og/utils/functionComposition.js [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    run: f => compose(f(x)), // runs the function 
+    runForEach: f => compose(x.map(el => f(el))), // runs the function for an array
+    end: () => x, // result
+})
+
+export {compose}
+
+// Example
+
+// const lower = str => str.toLowerCase();
```
Comment (Zemledelec) [inline src/og/utils/functionComposition.js]: clean up commented code

**Review by Zemledelec — CHANGES_REQUESTED** [review]

**Comment by Zemledelec** [comment]
> I have kept the old LayerSwitcher and name the new one **_LayerSwitcherv2_** Also this breaks functionality with other switchers. If you approve it I can change the other switchers too.

No need to use multiple switchers. Remove the first one.

File: src/og/control/LayerSwitcherv2.js (hunk 3 above)
Comment (martyrion) [reply src/og/control/LayerSwitcherv2.js]: Hey! I recall when you said maybe using $ to distinguish DOM elements from other variables. I used this convention. I will rename them though

File: src/og/control/LayerSwitcherv2.js (hunk 2 above)
Comment (martyrion) [reply src/og/control/LayerSwitcherv2.js]: Good, I will change that

File: css/og.css (hunk 1 above)
Comment (martyrion) [reply css/og.css]: I will check.

File: src/og/control/LayerSwitcherv2.js (hunk 4 above)
Comment (martyrion) [reply src/og/control/LayerSwitcherv2.js]: Good.

File: src/og/control/LayerSwitcherv2.js (hunk 5 above)
Comment (martyrion) [reply src/og/control/LayerSwitcherv2.js]: OK

File: src/og/control/LayerSwitcherv2.js (hunk 6 above)
Comment (martyrion) [reply src/og/control/LayerSwitcherv2.js]: Yes, I will fix it

File: src/og/control/LayerSwitcherv2.js (hunk 7 above)
Comment (martyrion) [reply src/og/control/LayerSwitcherv2.js]: Yes, noticed it also my self

File: src/og/control/LayerSwitcherv2.js (hunk 6 above)
Comment (martyrion) [reply src/og/control/LayerSwitcherv2.js]: I (think) it is fixed. I added a button to toggle activate/deactivate the switcher. For some reason after deactivation, I have to activate/deactivate/activate to get it to work again. I will look into it!

**Review by Zemledelec — APPROVED** [review]
Good @martyrion ! In advance, lets refactor huge methods. It's good practice making functions as small as possible.

**Comment by martyrion** [comment]
Hey! You are right, I will do so right away. What do you think about the options/ way of controlling the switcher? Is that good? Could it be applied to other switchers, or keep them separate as you fore-told?

**Comment by Zemledelec** [comment]
Let's move step by step, and make improvements on a solid base of shortage of functionality we meet. I suggest you provide a list of such options first, draw a picture, and then we will decide. Also, check an issue about Dialog refactoring, what do you think? As for me I always think a way of simplicity and efficiency. It means that the smaller - the better, the faster - the better, also always think about agile(SOLID principles) interfaces(code), it's the hardest part actually:)

**Pull request merged**